### DISCUSSION TITLE
Icon conventions

### DISCUSSION THREAD

**Discussion by jasongitmail** [sveltejs/kit#12648] [Ideas]
I'd like to see more conventions in SvelteKit.

Conventions 1.) help devs move faster, 2.) help the ecosystem grow faster by reducing waste researching/trying/rebuilding the same wheel, and 3.) **in this case, improve composability of components, which will become especially useful with AI generation and as more devs build re-usable component libraries (e.g. [Shadcn Svelte](https://www.shadcn-svelte.com/), [Aceternity Svelte](https://aceternity.sveltekit.io/), etc)** (to allow direct component usage, prevent duplicate loading of the same icon due to different icon loading methods, avoid requiring different setup, etc).

Icons are an opportunity for a convention (in official docs and create-svelte-app template) because they seem to be a solved problem given the below strategy (credit to Anthony Fu).

## Set up

1. Install dependencies: `npm add -D unocss @iconify/json`
2. Set up `vite.config.ts`

```ts
import { presetIcons } from "unocss";
import UnoCSS from "unocss/vite";

export default defineConfig({
  plugins: [
    UnoCSS({
      presets: [
        presetIcons({
          prefix: "i-",
          extraProperties: {
            display: "inline-block",
          }
        }),
      ],
    }),
    sveltekit(),
  ],
});
```
3. In `src/routes/+layout.svelte`, add:

```svelte
<script lang="ts">
  import "uno.css";
</script>
```

## Usage

```html
<div class="i-logos-svelte text-2xl"></div>
```

## Benefits of this icon approach

1.  Iconify provides a standard interface to icons
2. You can still side load icons not published on Iconify or customize icons.
3.  presetIcons converts SVG into CSS (so it doesn't increase JS bundle size like `unplug-icons` does)
4. perfect DX (no need to import anything before using an icon--just specify the icon where used).
5. Efficiency - the CSS containing each icon is included–at most once–sitewide in a far-futures cached CSS file. No unused icons are included.

## Background

- https://antfu.me/posts/icons-in-pure-css Anthony Fu explains what lead to this approach; he was using `unplug-icons` that he also created.
- https://unocss.dev/presets/icons

## The ask

I'd like to see this as a convention in SvelteKit templates and docs. It's easy to set up, once a person knows how, but having conventions provides important benefits noted in the intro.

CC @benmccann For his Twitter thread where I originally saw Anthony Fu's response.

**Comment by benmccann** [comment]
It sounds to me like you're asking for https://github.com/svelte-add/svelte-add/issues/315

**Reply by jasongitmail** [reply]
No. This misses the point.

It's easy to add. I did it already.

**I'm petitioning for the value of conventions in SvelteKit.**   

### Convention?

By "convention", I mean in the docs and ideally in the create-svelte-app template(s). 

### Who decides?

Whether it is this or another approach the core team feels is the most optimal solution known to date, for a given problem (icons, SEO, sitemap, etc); ideally open a Github issue to solicit other approaches if it an optimal approach is unclear.

### Why?

Lack of conventions imposes a cost on the community.

Although this only takes a few minutes to add, this isn't the cost to devs or the ecosystem, it's x1,000 devs doing the same things:

1. everyone researching, evaluating, building inferior solutions (like those I've in seen in Discord recently). Time that could have gone into their apps or advancing the Svelte ecosystem in more useful ways or iterating on the best known approach.
2. in this particular area, a convention would help component lib authors know how to handle icons in a more unified way, improving interoperability of components & page layouts created by various devs. 

Strong leadership in the area conventions is important to optimize DX ("optimize for vibes") and help the ecosystem grow.  

Whether is is this or something else, a documented convention would be good please, to help others.

**Reply by aloker** [reply]
It's great of you to try and move SvelteKit and is community forward. However, I don't think this should be baked into SvelteKit. Why UnoCSS? Why not Tailwind with the iconify plugin? Why not unplugin-icons? Those are user land decisions, not part of SvelteKit imo. I agree that developers should be taught about these solutions. Why not write a blog post about it or add it to https://github.com/one-aalam/awesome-svelte-kit or the likes?

**Reply by jasongitmail** [reply]
Not unocss as a whole, _the icon loading mechanism using vite_ from it or a similar approach.

> Why not write a blog post about it or add it to https://github.com/one-aalam/awesome-svelte-kit or the likes?

I addressed this in the Why section--same answers. Conventions provide loads of benefits, strategic ones too in this case; **people can deviate from a convention with much lower effort & time cost on their part, than the cost imposed on forcing the _entirety_ of all users to research/evaluate/implement an approach just to get going** and for them to still not know if it's considered the general best practice the team is aware of currently. Conventions allow for bubbling knowledge up about best known approaches.

**Reply by jasongitmail** [reply]
> Why not Tailwind with the iconify plugin? Why not unplugin-icons? 

The intent of your questions seems to be "who decides what is the best known approach?"
That is answered in my "who decides?" section. I.e. the core team based on what they know as best approach fitting the SvelteKit ethos currently, with community feedback to help bubble up ideas initially & iterate on the selected approach over time.

But if you're specifically interested in this topic, `unplug-icons` (answered in my OP too) is an older approach by the same author, Anthony Fu; it creates a JS file for each icon (increasing JS bundle size & parsing time). This approach converts the SVGs into CSS data (zero increase to JS size or parsing time) and includes them once in your global stylesheet automatically. So it's objectively better.

To illustrate the inefficiency of lacking a convention, whether this or another:
- Ben [tweeted](https://x.com/BenjaminMcCann/status/1831190151793639533) discovering `unplug-icons` yesterday. He's a maintainer that I have great respect for, and yet I as a mere mortal have been using `unplug-icons` for a year b/c I spent lots of time trying various approaches (first inefficiency in this story)
- Anthony Fu [responded](https://x.com/antfu7/status/1831258981051797621) to the tweet that there is an even better way--turns out he created this _years ago_! And neither of us were aware. (Second inefficiency b/c might have bubbled up earlier if a recommended approach were stated earlier)
- Now we've learned something, and the ability to bubble this up into a convention would help others.

**Reply by dominikg** [reply]
agree discovery of these options could be easier, but a category for "icons" in svelte-add would do that, combine it with talking about these more. Like I did a few svelte summits ago in a mini series called "sveltable vector graphics".

**Reply by jasongitmail** [reply]
_If discovery were the only goal_, a section on the official docs would be more effective. 

But that misses out on most the other benefits: devs will still be obligated to spend time evaluating different libs before choosing, it prevents sharing of the Svelte team's opinion on current best practice for SvelteKit to save others this time researching, inhibits knowledge bubbling about current known best practice, (in this case) prevents standardization around a given strategy that would help the component libs ecosystem grow in a compatible way, etc.

Lacking this, it relegates icons (& sitemap, SEO metadata, etc) to the same era of image plugins prior to enhanced:image, where half a dozen partially-broken solutions exist. 

**This doesn't mean everything needs to be core. It means the team would serve the community better by providing an opinion in the docs on the pieces that it is known to lack**, & in some cases, also including those pieces as part of the project templates. For common pieces.

**Reply by benmccann** [reply]
`create-svelte`/`svelte-add` is largely meant to be our recommendations of good defaults. That's why we offer vitest/playwright, but not other testing libraries. We've talked about adding docs for those CLIs to svelte.dev, which might help with discoverability via Google.

I wasn't aware of Tailwind with the Iconify plugin until now, but if I had to make a recommendation for something that's probably what it would be after having tried a few of these out now. I do like the CSS approach and unocss seems nice, but Tailwind is just way more popular than unocss.

`svelte-add` also offers a Tailwind integration. There's been some discussion of having it ask a question about plugins. Right now it just asks yes/no for typography, but we could instead make it a longer list of plugins with a checkbox for each of typography, forms, iconify, etc.

**Reply by aloker** [reply]
+1 for more tailwind plugins in svelte-add (including icobify)

**Reply by jasongitmail** [reply]
(Thanks everyone. I appreciate you hearing me out on the importance of conventions & I'll leave it at that. I'd welcome sharing other areas I have experienced in svelte spending time where I wish it hadn't been necessary and hope to help others can skip researching & rebuilding some of these wheels unless they choose to.) 

> I wasn't aware of Tailwind with the Iconify plugin until now, but if I had to make a recommendation for something that's probably what it would be after having tried a few of these out now. I do like the CSS approach a

Appreciate the opinion!

An ideal solution would: 
1. use iconify (standardizes interface to icons)
2. be performant (no JS icons)
3. be agnostic of the styling framework (for community benefits) 

Pros/Cons of `@iconify/tailwind`:
- 🟢 Uses Iconify 
- 🟢 Similar performance (SVG converted into CSS)
- ❌ Not agnostic of the styling framework b/c requires `tailwind.config.js`
- ❌ TailwindCSS v4 intends to move to a CSS-based configuration by default; this would lock users into the JS config.
- ❌ Weird syntax (for "dynamic" font family imports):

    ```html
    <!-- using @iconify/tailwind -->
    <span class="icon-[mdi-light--home]"></span>

    <!-- using @egoist/tailwindcss-icons -->
    <span class="i-[mdi-light--home]"></span>

- ❌🟢 weird syntax (for specific icon family imports), but egoist's is good: 

    ```html
    <!-- Usage with @iconify/tailwind -->
    <span class="iconify mdi-light--home"></span>
    <span class="iconify-color vscode-icons--file-type-tailwind"></span>

    <!-- Usage with @egoist/tailwindcss-icons -->
    <span class="i-mdi-light-home"></span>
    ```
    
 I still see unocss' version as more optimal for SK (primarily b/c it's a Vite plugin):
   
- 🟢 Uses Iconify 
- 🟢 Similar performance (SVG converted into CSS)
- 🟢 Agnostic of the styling framework in the sense that it's a Vite plugin, ❌ although requires importing `uno.css` so not as ideal as it could be. almost!
- 🟢 Nice syntax (`i-mdi-light-home` or `i-mdi:light-home`)

**Comment by dominikg** [comment]
> I wasn't aware of Tailwind with the Iconify plugin until now, but if I had to make a recommendation for something that's probably what it would be after having tried a few of these out now. I do like the CSS approach and unocss seems nice, but Tailwind is just way more popular than unocss.

at least tailwind v3 is technically inferior to unocss in vite and might be harder to integrate because you need a custom extractor for it.
